include/sys/ddt.h

```diff
@@ -120,60 +127,149 @@ typedef struct {
  * we maintain the ability to free existing dedup-ditto blocks.
  */
-enum ddt_phys_type {
+
+typedef enum {
 	DDT_PHYS_DITTO = 0,
 	DDT_PHYS_SINGLE = 1,
 	DDT_PHYS_DOUBLE = 2,
 	DDT_PHYS_TRIPLE = 3,
-	DDT_PHYS_TYPES
-};
+	DDT_PHYS_FLAT = 4,
+	DDT_PHYS_NONE = 5
+} ddt_phys_variant_t;
+
+#define	DDT_PHYS_VARIANT(ddt, p)	\
+	(ASSERT((p) < DDT_PHYS_NONE),	\
+	((ddt)->ddt_flags & DDT_FLAG_FLAT ? DDT_PHYS_FLAT : (p)))
+
+#define	DDT_TRAD_PHYS_SIZE	sizeof (((ddt_univ_phys_t *)0)->ddp_trad)
+#define	DDT_FLAT_PHYS_SIZE	sizeof (((ddt_univ_phys_t *)0)->ddp_flat)
+
+#define	_DDT_PHYS_SWITCH(ddt, flat, trad)	\
+	(((ddt)->ddt_flags & DDT_FLAG_FLAT) ? (flat) : (trad))
+
+#define	DDT_PHYS_SIZE(ddt)		_DDT_PHYS_SWITCH(ddt,	\
+	DDT_FLAT_PHYS_SIZE, DDT_TRAD_PHYS_SIZE)
+
+#define	DDT_NPHYS(ddt)			_DDT_PHYS_SWITCH(ddt, 1, DDT_PHYS_MAX)
+#define	DDT_PHYS_FOR_COPIES(ddt, p)	_DDT_PHYS_SWITCH(ddt, 0, p)
+#define	DDT_PHYS_IS_DITTO(ddt, p)	_DDT_PHYS_SWITCH(ddt, 0, (p == 0))
 
 /*
  * A "live" entry, holding changes to an entry made this txg, and other data to
  * support loading, updating and repairing the entry.
  */
 
 /* State flags for dde_flags */
-#define	DDE_FLAG_LOADED (1 << 0)	/* entry ready for use */
+#define	DDE_FLAG_LOADED		(1 << 0)	/* entry ready for use */
+#define	DDE_FLAG_OVERQUOTA	(1 << 1)	/* entry unusable, no space */
+#define	DDE_FLAG_LOGGED		(1 << 2)	/* loaded from log */
 
+/*
+ * Additional data to support entry update or repair. This is fixed size
+ * because its relatively rarely used.
+ */
 typedef struct {
-	/* key must be first for ddt_key_compare */
-	ddt_key_t	dde_key;			/* ddt_tree key */
-	ddt_phys_t	dde_phys[DDT_PHYS_TYPES];	/* on-disk data */
+	/* copy of data after a repair read, to be rewritten */
+	abd_t		*dde_repair_abd;
+
+	/* original phys contents before update, for error handling */
+	ddt_univ_phys_t	dde_orig_phys;
 
 	/* in-flight update IOs */
-	zio_t		*dde_lead_zio[DDT_PHYS_TYPES];
+	zio_t		*dde_lead_zio[DDT_PHYS_MAX];
+} ddt_entry_io_t;
 
-	/* copy of data after a repair read, to be rewritten */
-	struct abd	*dde_repair_abd;
+typedef struct {
+	/* key must be first for ddt_key_compare */
+	ddt_key_t	dde_key;	/* ddt_tree key */
+	avl_node_t	dde_node;	/* ddt_tree_node */
 
 	/* storage type and class the entry was loaded from */
 	ddt_type_t	dde_type;
 	ddt_class_t	dde_class;
 
 	uint8_t		dde_flags;	/* load state flags */
 	kcondvar_t	dde_cv;		/* signaled when load completes */
+	uint64_t	dde_waiters;	/* count of waiters on dde_cv */
 
-	avl_node_t	dde_node;	/* ddt_tree node */
+	ddt_entry_io_t	*dde_io;	/* IO support, when required */
+
+	ddt_univ_phys_t	dde_phys[];	/* flexible -- allocated size varies */
 } ddt_entry_t;
 
+/*
+ * A lightweight entry is for short-lived or transient uses, like iterating or
+ * inspecting, when you don't care where it came from.
+ */
+typedef struct {
+	ddt_key_t	ddlwe_key;
+	ddt_type_t	ddlwe_type;
+	ddt_class_t	ddlwe_class;
+	uint8_t		ddlwe_nphys;
```

Comment:
I've noticed that in all 4 places where `ddlwe_nphys` is used we also have ddt.  We could save 8 bytes of ddt_lightweight_entry_t size by using `DDT_NPHYS(ddt)` instead, unless you think this abstraction really has its use, which does not look clean after addition of `ddp_class_start`, that makes `ddp_flat` more than 1/4th of `ddp_trad`.

Reply:
Yeah, that's fair, removed on the `fdt-rel-flat` branch.

 It made more sense in the earliest conception of lightweight entry, where it was totally self-contained and didn't need a `ddt_t` to interpret it, and that was useful in a few places where it was just oeprating on some number of `ddt_phys_t`.  But right now as you note, that doesn't hold anymore, and there's always the `ddt` there that we need, so we might as well use it.

It could probably be made a bit less weird now, I suppose. eg, `ddt_univ_phys_t` could maybe carry a trad/flat flag, or maybe `ddt_walk` could be returning the `ddt_t` as well (or that's in the entry too). Or maybe even a rework of `ddt_entry_t` and `ddt_log_entry_t` to have a consistent internal structure such that the copy isn't necessary. I'm not going to attempt it yet though; another thing that I think we can look at once this is all settled.